server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java

```diff
@@ -214,44 +214,117 @@ public BytesReference build() throws IOException {
+
+            // NOTE: catenate all dimension value hashes up to names certain number of dimensions
+            int numberOfDimensions = Math.min(MAX_DIMENSIONS, dimensions.size());
+            int buf1Index = 0;
+            final byte[] buf1 = new byte[4 * numberOfDimensions];
+            for (final DimensionDataHolder dimension : dimensions) {
+                if (buf1Index >= buf1.length) break;
+                final BytesRef dimensionValueBytesRef = dimension.value().toBytesRef();
+                ByteUtils.writeIntLE(StringHelper.murmurhash3_x86_32(dimensionValueBytesRef, 0), buf1, buf1Index);
+                buf1Index += 4;
+            }
+            final BytesRef valueByValue = new BytesRef(buf1, 0, buf1.length);
+
+            // NOTE: hash all dimension field allValues
+            int buf2Index = 0;
+            final byte[] buf2 = new byte[dimensionValuesLength];
+            for (final DimensionDataHolder dimension : dimensions) {
+                final BytesRef dimensionValueBytesRef = dimension.value.toBytesRef();
+                System.arraycopy(
+                    dimensionValueBytesRef.bytes,
+                    dimensionValueBytesRef.offset,
+                    buf2,
+                    buf2Index,
+                    dimensionValueBytesRef.length - dimensionValueBytesRef.offset
+                );
+                buf2Index += (dimensionValueBytesRef.length - dimensionValueBytesRef.offset);
+            }
+            final BytesReference dimensionValuesHash = hash128(new BytesRef(buf2, 0, buf2.length));
+
+            final BytesRef names = dimensionNamesHash.toBytesRef();
+            final BytesRef metrics = metricFieldsHash.toBytesRef();
+            final BytesRef allValues = dimensionValuesHash.toBytesRef();
+            int tsidLength = names.length + metrics.length + valueByValue.length + allValues.length;
+            byte[] tsid = new byte[tsidLength];
+            int tsidIndex = 0;
+            System.arraycopy(names.bytes, names.offset, tsid, tsidIndex, names.length);
+            tsidIndex += names.length;
+            System.arraycopy(metrics.bytes, metrics.offset, tsid, tsidIndex, metrics.length);
+            tsidIndex += metrics.length;
+            System.arraycopy(valueByValue.bytes, valueByValue.offset, tsid, tsidIndex, valueByValue.length);
+            tsidIndex += valueByValue.length;
+            System.arraycopy(allValues.bytes, allValues.offset, tsid, tsidIndex, allValues.length);
+            tsidIndex += allValues.length;
+            assert tsidIndex == tsidLength;
+            try (BytesStreamOutput out = new BytesStreamOutput()) {
+                out.writeVInt(TSID_HASH_SENTINEL);
+                out.writeBytesRef(new BytesRef(tsid, 0, tsidLength));
+                return out.bytes();
+            }
+        }
+
+        private BytesReference hash128(final BytesRef value) throws IOException {
```

Comment:
I think there's some potential for optimization by providing a pre-allocated `java.nio.ByteBuffer` as a parameter of the `hash128` method here that gets initialized in the beginning of `similarityHash`. We know how large the `ByteBuffer` will need to be based on the number of dimensions: `16 * 3 + Math.min(MAX_DIMENSIONS, dimensions.size())`.

We can then write into the pre-allocated buffer like this:
```java
buffer.putLong(hash128.h1);
buffer.putLong(hash128.h2);
```
instead of returning a `BytesReference`.

This avoids allocating another `byte[]`, a `BytesStreamOutput`, including  the internal `ByteArray` buffer, and a `BytesRef`.

Reply:
I changed that to use just one byte buffer in a followup pr